services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/services/generic_scheduler/_store.py

```diff
@@ -0,0 +1,432 @@
+    return json_loads(obj_str)
+
+
+class Store:
+    def __init__(self, redis_settings: RedisSettings) -> None:
+        self.redis_settings = redis_settings
+
+        self._client: RedisClientSDK | None = None
+
+    async def setup(self) -> None:
+        self._client = RedisClientSDK(
+            self.redis_settings.build_redis_dsn(RedisDatabase.DYNAMIC_SERVICES),
+            client_name=__name__,
+        )
+        await self._client.setup()
+
+    async def shutdown(self) -> None:
+        if self._client:
+            await self._client.shutdown()
+
+    @property
+    def redis(self) -> aioredis.Redis:
+        assert self._client  # nosec
+        return self._client.redis
+
+    async def set_multiple(self, hash_key: str, updates: dict[str, Any]) -> None:
+        """saves multiple key-value pairs in a hash"""
+        await handle_redis_returns_union_types(
+            self.redis.hset(
+                hash_key, mapping={k: _dumps(v) for k, v in updates.items()}
+            )
+        )
+
+    async def set(self, hash_key: str, key: str, value: Any) -> None:
+        """saves a single key-value pair in a hash"""
+        await self.set_multiple(hash_key, {key: value})
+
+    async def get(self, hash_key: str, *keys: str) -> tuple[Any, ...]:
+        """retrieves one or more keys from a hash"""
+        result: list[str | None] = await handle_redis_returns_union_types(
+            self.redis.hmget(hash_key, list(keys))
+        )
+        return tuple(_loads(x) if x else None for x in result)
+
+    async def delete(self, hash_key: str, *keys: str) -> None:
+        """removes one or more keys form a hash"""
+        await handle_redis_returns_union_types(self.redis.hdel(hash_key, *keys))
+
+    async def remove(self, *hash_keys: str) -> None:
+        """removes the entire hash"""
+        await handle_redis_returns_union_types(self.redis.delete(*hash_keys))
+
+    async def increase_and_get(self, hash_key: str, key: str) -> NonNegativeInt:
+        """increasea a key in a hash by 1 and returns the new value"""
+        return await handle_redis_returns_union_types(
+            self.redis.hincrby(hash_key, key, amount=1)
+        )
+
+    async def decrease_and_get(self, hash_key: str, key: str) -> NonNegativeInt:
+        """decrease a key in a hash by 1 and returns the new value"""
+        return await handle_redis_returns_union_types(
+            self.redis.hincrby(hash_key, key, amount=-1)
+        )
+
+
+class _UpdateScheduleDataDict(TypedDict):
+    operation_name: NotRequired[OperationName]
+    group_index: NotRequired[NonNegativeInt]
+    is_creating: NotRequired[bool]
+    operation_error_type: NotRequired[OperationErrorType]
+    operation_error_message: NotRequired[str]
+
+
+_DeleteScheduleDataKeys = Literal[
+    "operation_name",
+    "group_index",
+    "is_creating",
+    "operation_error_type",
+    "operation_error_message",
+]
+
+
+class ScheduleDataStoreProxy:
```

Comment:
Q: why proxy?

I understand that this object is used to CRUD operations on the keys in `_DeleteScheduleDataKeys`